### PR TITLE
D1 api auth

### PR DESCRIPTION
**Pull request by shubaivan**
auth action

### CODE REVIEW

File: src/AppBundle/Controller/Api/UserController.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+class UserController extends Controller
+{
+    /**
+     * @Route("/api/authorization", name="api_projects_list")
+     * @Method({"GET", "POST"})
+     */
+    public function authAction(Request $request)
+    {
+//        $code = $request->query->get('code');
```
Comment (ahonymous) [inline src/AppBundle/Controller/Api/UserController.php]: $code = $request->query->get('code');

File: src/AppBundle/Controller/Api/UserController.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Method;
+use Symfony\Bundle\FrameworkBundle\Controller\Controller;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\Routing\Annotation\Route;
+use Symfony\Component\HttpFoundation\Request;
+
+class UserController extends Controller
+{
+    /**
+     * @Route("/api/authorization", name="api_projects_list")
```
Comment (ahonymous) [inline src/AppBundle/Controller/Api/UserController.php]: name of route should be "api_authorization"

File: src/AppBundle/Controller/Api/UserController.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+     * @Method({"GET", "POST"})
+     */
+    public function authAction(Request $request)
```
Comment (ahonymous) [inline src/AppBundle/Controller/Api/UserController.php]: authorizationAction

File: src/AppBundle/Security/BankIdService.php [hunk 4]
```diff
@@ -46,7 +46,8 @@ public function getAccessToken($code)
 
         $accessToken = $client->get($url)->send()->getBody(true);
 
-        $this->getBankIdUser($accessToken);
```
Comment (ahonymous) [inline src/AppBundle/Security/BankIdService.php:49]: return $this->getBankIdUser($accessToken);

File: src/AppBundle/Security/BankIdService.php [hunk 5]
```diff
@@ -46,7 +46,8 @@ public function getAccessToken($code)
...
+        $data = $this->getBankIdUser($accessToken);
```
Comment (ahonymous) [inline src/AppBundle/Security/BankIdService.php]: remove

File: src/AppBundle/Security/BankIdService.php [hunk 6]
```diff
@@ -46,7 +46,8 @@ public function getAccessToken($code)
...
+        return $data;
```
Comment (ahonymous) [inline src/AppBundle/Security/BankIdService.php]: remove

File: src/AppBundle/Controller/Api/UserController.php (hunk 1 above)
Comment (shubaivan) [inline src/AppBundle/Controller/Api/UserController.php]: done, I use hard code for debug

File: src/AppBundle/Controller/Api/UserController.php (hunk 2 above)
Comment (shubaivan) [inline src/AppBundle/Controller/Api/UserController.php]: done

File: src/AppBundle/Controller/Api/UserController.php (hunk 3 above)
Comment (shubaivan) [inline src/AppBundle/Controller/Api/UserController.php]: done

File: src/AppBundle/Security/BankIdService.php (hunk 4 above)
Comment (shubaivan) [inline src/AppBundle/Security/BankIdService.php:49]: done

File: src/AppBundle/Helper/UserManager.php [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    protected $em;
+
+    private $rootDir;
+
+    public function setRootDir($rootDir)
+    {
+        $this->rootDir = $rootDir;
+    }
+
+    public function __construct($_templateEngine, EntityManager $em)
```
Comment (ahonymous) [inline src/AppBundle/Helper/UserManager.php:21]: What you render with template engine?

File: src/AppBundle/Security/BankIdService.php [hunk 8]
```diff
@@ -34,6 +34,7 @@ public function getLink()
 
     public function getAccessToken($code)
     {
+//        dump($code);exit;
```
Comment (ahonymous) [inline src/AppBundle/Security/BankIdService.php:37]: remove this

File: src/AppBundle/Helper/UserManager.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        $this->templateEngine = $_templateEngine;
+        $this->em = $em;
+    }
+
+    public function isUniqueUser($clid)
+    {
+        $found = $this->em->getRepository('AppBundle:User')->findByClid($clid);
+        if (empty($found)) {
```
Comment (ahonymous) [inline src/AppBundle/Helper/UserManager.php]: if (!$user) {
    $user = new User();
    ....
}

File: src/AppBundle/Helper/UserManager.php [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+            try {
+                //ToDo create new User
+//                $this->em->flush();
+//                $this->mail->toUser_registration($user);//congratulation for registration
+            } catch (Exception $e) { return false; }
+        } else {
```
Comment (ahonymous) [inline src/AppBundle/Helper/UserManager.php]: remove else

File: src/AppBundle/Helper/UserManager.php [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+            return false;
```
Comment (ahonymous) [inline src/AppBundle/Helper/UserManager.php]: return User

File: src/AppBundle/Helper/UserManager.php [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+        }
+        return $found;
+    }
+
+//    public function getInfoIpCountry($ip)
```
Comment (ahonymous) [inline src/AppBundle/Helper/UserManager.php]: remove comments code

File: src/AppBundle/Resources/config/services.yml [hunk 13]
```diff
@@ -15,3 +15,10 @@ services:
             - @app.security.bank_id
         tags:
             - { name: twig.extension }
+
+    app.user.manager:
+        class:  AppBundle\Helper\UserManager
+        arguments: [@templating, @doctrine.orm.entity_manager ]
+
+        calls:
+            - [ setRootDir, ["%kernel.root_dir%"] ]
```
Comment (ahonymous) [inline src/AppBundle/Resources/config/services.yml:24]: for what you injected root directory

**Comment by red-p1xel** [comment]
Alternative method realpath($this->container->getParameter('kernel.root_dir'))

File: src/AppBundle/Helper/UserManager.php (hunk 7 above)
Comment (shubaivan) [inline src/AppBundle/Helper/UserManager.php:21]: for future for find city and country by ip with help GeoLite. Like this $this->rootDir.'/data/GeoLite2-City.mmdb

File: src/AppBundle/Security/BankIdService.php (hunk 8 above)
Comment (shubaivan) [inline src/AppBundle/Security/BankIdService.php:37]: done

File: src/AppBundle/Helper/UserManager.php (hunk 9 above)
Comment (shubaivan) [inline src/AppBundle/Helper/UserManager.php]: I thunk, if found empty(not find user by Clid) service create new user

File: src/AppBundle/Helper/UserManager.php (hunk 11 above)
Comment (shubaivan) [inline src/AppBundle/Helper/UserManager.php]: done

File: src/AppBundle/Resources/config/services.yml (hunk 13 above)
Comment (shubaivan) [inline src/AppBundle/Resources/config/services.yml:24]: for future, for GeoLite

Comment (ahonymous) [inline src/AppBundle/Resources/config/services.yml:24]: I asked "for what", not "How".

File: src/AppBundle/Helper/UserManager.php (hunk 7 above)
Comment (spolischook) [inline src/AppBundle/Helper/UserManager.php:21]: You not need template engine for "find city and country by ip"!

File: src/AppBundle/Helper/UserManager.php [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+    public function setRootDir($rootDir)
+    {
+        $this->rootDir = $rootDir;
+    }
+
+    public function __construct($_templateEngine, EntityManager $em)
+    {
+        $this->templateEngine = $_templateEngine;
+        $this->em = $em;
```
Comment (spolischook) [inline src/AppBundle/Helper/UserManager.php:24]: Inject Repository instead EM

File: src/AppBundle/Resources/config/services.yml (hunk 13 above)
Comment (spolischook) [inline src/AppBundle/Resources/config/services.yml:24]: in future you can inject, don't do anything if it not related to this PR. KISS.

File: src/AppBundle/Security/BankIdService.php (hunk 4 above)
Comment (spolischook) [inline src/AppBundle/Security/BankIdService.php:49]: You need to add doc block for methods. 
Why method getAccessToken return BankIdUser?

File: src/AppBundle/Controller/Api/UserController.php [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+class UserController extends Controller
+{
+    /**
+     * @Route("/api/authorization", name="api_authorization")
+     * @Method({"GET", "POST"})
+     */
+    public function authorizationAction(Request $request)
+    {
+//        $code = $request->query->get('code');
+        $code = 'N4iFZw50015';
```
Comment (spolischook) [inline src/AppBundle/Controller/Api/UserController.php]: use it for tests, not for code

File: src/AppBundle/Entity/User.php [hunk 16]
```diff
@@ -13,7 +13,7 @@
  * @ORM\Table()
  * @ORM\Entity(repositoryClass="AppBundle\Entity\UserRepository")
  */
-class User implements UserInterface
+class User implements UserInterface, \JsonSerializable
```
Comment (spolischook) [inline src/AppBundle/Entity/User.php:17]: I guess JsonSerializable Interfase mean that all data in class can be represented as a json, not just clid.

File: src/AppBundle/Helper/UserManager.php (hunk 7 above)
Comment (spolischook) [inline src/AppBundle/Helper/UserManager.php:21]: YAGNI

File: src/AppBundle/Resources/config/services.yml (hunk 13 above)
Comment (spolischook) [inline src/AppBundle/Resources/config/services.yml:24]: YAGNI

**Comment by spolischook** [comment]
The first thing that you need to do, when you write security or money - tests

File: src/AppBundle/Security/BankIdService.php (hunk 4 above)
Comment (shubaivan) [inline src/AppBundle/Security/BankIdService.php:49]: return $this->getBankIdUser($accessToken); have information from bank about user

File: src/AppBundle/Helper/UserManager.php (hunk 7 above)
Comment (shubaivan) [inline src/AppBundle/Helper/UserManager.php:21]: maybe use template engine for future for send email congratulation user for registration

File: src/AppBundle/Entity/User.php (hunk 16 above)
Comment (shubaivan) [inline src/AppBundle/Entity/User.php:17]: now return getFullName = return $this->getLastName() . ' ' . $this->getFirstName() . ' ' . $this->getMiddleName(); I don’t know what exactly return needed

**Pull request closed without merging**